### ISSUE TITLE
ENH: Support free-threaded python build (tracking issue)

### ISSUE THREAD

**Issue by ngoldbaum** [numpy/numpy#26157]
Myself, @lysnikolaou, and @rgommers are working on a new project in collaboration with the Python Runtime team at Meta to add support for Python 3.13 nogil builds in the scientific python software stack.

Our initial minimal goal is a build of NumPy that runs and passes all the tests in a nogil build of Python.

Resources:

* [PEP 703](https://peps.python.org/pep-0703/)
* [Fork of numpy 1.24 with early nogil support](https://github.com/numpy/numpy/compare/v1.24.0...colesbury:numpy:v1.24.0-nogil)

* [x] Add changelog entry and docs describing the current state of using NumPy in a multithreaded environment.
    * https://github.com/numpy/numpy/pull/27200
* [x] Convert `PyDict_GetItem` and `PyList_GetItem` to use variants returning strong references (https://github.com/numpy/numpy/issues/26159)
    * [x] Vendor [pythoncapi-compat](https://github.com/python/pythoncapi-compat) in a submodule (https://github.com/numpy/numpy/issues/26158)
* [x] #26161
    * #26190
    * https://github.com/numpy/numpy/pull/26207
* [x] Set up CPython 3.13 CI
    * [x] 3.13 gil enabled build (https://github.com/numpy/numpy/pull/26284)
    * [x] `--disable-gil` build (#26463)
* [x] Make the test suite pass with the gil disabled
* [x] Mark NumPy C extensions with `Py_MOD_GIL_NOT_USED`.
    * https://github.com/numpy/numpy/pull/26913
    * https://github.com/numpy/numpy/pull/26950
* [x] Allow building f2py extensions that can be imported without the GIL.
    * https://github.com/numpy/numpy/pull/26981
* [x] Use PyMutex instead of PyThread_type_lock
    * https://github.com/numpy/numpy/pull/27011
* [x] Nightly wheels for `cp313t` ABI
    * [x] manylinux and musllinux #26512 
    * [x] MacOS #26664
    * [x] Windows #27560
* [x] Global state
    * C static variables in `_multiarray_umath` reorganized into structs: https://github.com/numpy/numpy/pull/26607  
    * Global caches need to be disabled, have locking, serialize creating read-only caches, or converted to thread-local caches.
        *  Disabled:
            * #26251
            * #26294
            * #26348 
        * Locking:
            *  #26348 
            * https://github.com/numpy/numpy/pull/26430
            * https://github.com/numpy/numpy/pull/26750
        * Cache initialization moved to module initialization
            * https://github.com/numpy/numpy/pull/26607 
        * Some runtime caches need a single-init API. https://github.com/numpy/numpy/pull/26780
        * Dragon4 scientific printing code uses a static global scratch space.
        * Triaged low priority:
            * `wrapping_array_method.c` uses a freelist that is certainly not thread safe. It has no internal consumers besides `_scaled_float_dtype` to test the implementation so ignoring it for now.
            * the number of legacy user dtypes numpy knows about is held in a global static variable: `NPY_NUMUSERTYPES`, which also unfortunately is part of the public C API. This one may just need to remain unfixed, given it's in a legacy API.
            * The check to force an error on module reloading isn't thread safe. Probably needs an atomic to guard initializing it.
    * Global settings and state may need to be made per-thread or synchronized across threads.
        * Converted to thread-local state:
            *  https://github.com/numpy/numpy/pull/26368
            * https://github.com/numpy/numpy/pull/26676
        * `np.set_string_function` and `PyArray_SetStringFunction` are implemented using static global state.
            * These are removed, see #26576, https://github.com/numpy/numpy/pull/26611
        * Make printoptions a context variable. https://github.com/numpy/numpy/pull/26846
        * [hugepage support](https://numpy.org/doc/stable/reference/global_state.html#madvise-hugepage-on-linux) is controlled by a single static global. The only public interface to control this is an environment variable read at startup, so this isn't problematic unless someone is using private APIs. 
        * `NUMPY_WARN_IF_NO_MEMORY_POLICY` is implemented using a global static. The only public interface is via  an environment variable read at module initialization but it's tested using a thread-unsafe interface.

**Comment by ogrisel** [comment]
Maybe another interesting resource is the past work done by Sam Gross to patch numpy to make it work well enough with the original `nogil` fork of CPython 3.9.

https://github.com/numpy/numpy/compare/v1.24.0...colesbury:numpy:v1.24.0-nogil

Things might have drifted a bit since that time, but those patches are enough to get the scikit-learn test suite pass with `nogil`-CPython 3.9: see e.g. [this run](https://dev.azure.com/scikit-learn/scikit-learn/_build?definitionId=5&_a=summary&branchFilter=3834%2C3834%2C3834%2C3834) of the `Linux_nogil pylatest_pip_nogil` job that is scheduled run on a nightly basis on our Azure Pipelines CI.

**Comment by ngoldbaum** [comment]
It looks like numpy's main branch passes the full numpy test suite with the gil disabled. I've been testing using the `nogil-integration` branch in @colesbury's fork of CPython, but i tried just now using the CPython `main` branch and the test suite passes with the GIL disabled there as well.

There are still a ton of threading bugs I'm sure because of limited use of threads in the tests - we need better test coverage for threaded workflows. I'm already aware of a number of internal caches we'll likely need to disable or add locking for.

But, it's a good start!

**Comment by a-reich** [comment]
Just curious, are contributors planning to work on https://github.com/numpy/numpy/issues/24755 or is there any overlap where work for nogil would also help with allowing subinterpreters?

**Comment by ngoldbaum** [comment]
While this does have some overlap with work on subinterpreters since the C thread safety issues we're working on apply to subinterpreters as well, the bulk of the work required to support subinterpreters - convering numpy's extension types from single-phase to multi-phase initialization - is not required and we are not currently planning to do that.

**Comment by ngoldbaum** [comment]
There are now `cp313t` nightly wheels up for `manylinux` and `musllinux` at our normal spot for nightlies. Windows wheels will need to wait on resolving some windows-specific build issues and Mac wheels will need to wait for 3.13.0b2, which will have mac binaries cibuildwheel can use.

To install the wheel on a free-threaded build of python you'll need the prerelease of pip 24.1, installable with e.g. `pip install -U --pre pip`. The current stable release doesn't know about the new ABI tag and will install the wheel for the `cp313` ABI and then fail to import at runtime because of the ABI mismatch.

**Comment by rgommers** [comment]
> `lapack_lite` uses static globals all over the place and is not thread safe. Maybe we should require a real BLAS on the free-threaded build?

I think we should do that, at least for now. It may not be tenable in the very long term, but for the coming 1-2 years at least, everyone who is building for free-threaded CPython should know how to install a BLAS library.

There's no dedicated method for "is this a free-threaded build" yet in Meson. Something like this should do it though I think (untested):
```diff
diff --git a/numpy/meson.build b/numpy/meson.build
index 7e9ec5244c..ee9450ddc5 100644
--- a/numpy/meson.build
+++ b/numpy/meson.build
@@ -56,6 +56,11 @@ endif
 blas_name = get_option('blas')
 lapack_name = get_option('lapack')
 allow_noblas = get_option('allow-noblas')
+if py.get_variable('Py_GIL_DISABLED', fallback: '0') == '1'
+  # We don't allow using `lapack_lite` in the free-threaded build, it's unsafe (see gh-26157)
+  allow_noblas = false
+endif
+
 # This is currently injected directly into CFLAGS/CXXFLAGS for wheel builds
 # (see cibuildwheel settings in pyproject.toml), but used by CI jobs already
 blas_symbol_suffix = get_option('blas-symbol-suffix')
```

**Comment by rgommers** [comment]
As for global state, it'd be good to check `numpy.random`. There are nice parallel features for users that know what they are doing (see [docs](https://numpy.org/devdocs/reference/random/parallel.html)), but the worry is more about single-threaded-like usage:

- I'm not sure that it's safe to call create a `Generator` instance, create threads, then call methods from multiple threads when there is no GIL. There is pretty careful lock usage in `random/_generator.pyx` and calls that advance the bitgenerator state look atomic, but how sure are we that that's actually the case?
- The legacy API (`np.random.*` functions and `RandomState`) is implemented in the same fashion and so uses locks as well, but probably there are some separate concerns (e.g., calling `seed()` from one thread while generating numbers in another).

@rkern do you happen to know off the top of your head what the state is?

**Comment by rkern** [comment]
Not sure. GILlessness didn't exist at the time of the design, so it surely didn't enter our mental model when thinking about it. We did want to support `with nogil:` blocks that just call the C functions that draw data from and advance the `BitGenerator` state, so that is where the careful lock usage comes from. I don't know if that's going to be sufficient for other removals of the GIL where we were expecting there to be a GIL in place in the past.

**Comment by rkern** [comment]
And I haven't gotten my head around free-threaded Python yet to have a mental model to reason about it at this time.

**Comment by ngoldbaum** [comment]
I'll try to take a look with an eye toward thread safety. I'm not very familiar with the RNG internals so it'll be a good excuse to poke around. I might ping the mailing list or open an issue if I have questions to give @rkern an aswer to help clarify things :)

**Comment by ngoldbaum** [comment]
I looked at the `np.random` cython code today and I think we're fine without any new major updates.

All existing usages of the lowlevel bitgen state are guarded by acquiring a lock. We can probably make all of this faster using `PyThread_type_lock` directly or when it's available, using `PyMutex`.

The `SeedSequence` state is generated inside `__init__`, so no need to lock there because `__init__` will only ever be called on a single thread for a single instance. Any data races are also present with the GIL and require doing silly things like spawning child seed sequences from the same parent sequence in multiple threads.

**Comment by aaronzo** [comment]
Are there plans to release free-threaded builds of numpy 1.x? Or will this effort be strictly 2.x

**Comment by ngoldbaum** [comment]
No, there is no plan to do that, the first version of NumPy to support Python 3.13 will either be a 2.0.x version or NumPy 2.1 and the first version to support the free-threaded build with be NumPy 2.1.

**Comment by effigies** [comment]
Creating a free-threaded CI job over at https://github.com/nipy/nibabel/pull/1339.

Note that the GIL is being enabled by default:

```
<frozen importlib._bootstrap>:488: RuntimeWarning: The global interpreter lock (GIL) has been enabled to load module 'numpy._core._multiarray_umath', which has not declared that it can run safely without the GIL. To override this behavior and keep the GIL disabled (at your own risk), run with PYTHON_GIL=0 or -Xgil=0.
```

Disabling it with `PYTHON_GIL=0` passes our test suite, but figured I should note it here in case this isn't a known issue. We're installing https://pypi.anaconda.org/scientific-python-nightly-wheels/simple/numpy/2.1.0.dev0/numpy-2.1.0.dev0-cp313-cp313t-manylinux_2_17_x86_64.manylinux2014_x86_64.whl, uploaded 2024-07-10.

**Comment by ngoldbaum** [comment]
Thanks! That's a known issue. Cython doesn't have a way to add the `Py_mod_gil` slot to the extension. Hopefully cython nightly grows that feature but if worst comes to worst we might end up using a patched cython to build the numpy wheels we upload to pypi. For now you'll need to set the environment variable or pass `-Xgil=0`.

**Comment by rgommers** [comment]
@ngoldbaum it looks like that Cython PR will land in 1-2 weeks. Using that is only a single flag when we're invoking `cython` though, while converting the other extension modules outside of `numpy/random/` (which don't use Cython) is a larger diff. It'd be great to start using `Py_mod_gil` in all of those now, so (a) those extension modules get exercised even without `-Xgil=0` and (b) when we backport the Cython change to 2.1.x the diff will be small.

**Comment by ngoldbaum** [comment]
Good point, see https://github.com/numpy/numpy/pull/26913

**Comment by seberg** [comment]
Since its the tracking issue, mentioning here also:  Object arrays should not be fully thread safe since there is no locking when they are mutated  So one thread can decref before another incref's the same element (or am I missing some magic?).
There are approaches to this, may need to discuss in a separate issue eventually (maybe there is also some simple solution I don't know yet).

**Comment by ngoldbaum** [comment]
The Cython PR just got merged so I'll send in the necessary Cython build changes as well whenever the Cython wheel gets regenerated.

**Comment by rgommers** [comment]
> The Cython PR just got merged so I'll send in the necessary Cython build changes as well whenever the Cython wheel gets regenerated.

In case it helps: you cannot add the flag unconditionally and whether the interpreter is a free-threaded one cannot be (easily) detected in `meson.build`, so the way to do this is to add the flag unconditionally for Cython `>=3.1.0a0`. There's already a Cython version check in the top-level `meson.build` file.

**Comment by lysnikolaou** [comment]
The flag is okay to be added to Cython modules even under a normal (non-free-threaded) build, where it'll basically do nothing. So, yes, the way to go is add the flag unconditionally the Cython version that supports it.

**Comment by ngoldbaum** [comment]
I chatted with @colesbury about locking for object arrays today. He thinks we will need some sort of locking or synchronization to fix the threaded object array access issues @seberg is worried about.

Rather than adding a new lock, he suggested using the [critical section API](https://docs.python.org/3.13/c-api/init.html#python-critical-section-api) on the array object before and after spots we release or acquire the GIL for dtypes that don't need the GIL.

There are quite a few places we use the macros that release and acquire the GIL, but I also think it should be relatively straightforward to grep for those macros and find spots where we have a special check for object arrays so that we don't release the GIL in that case. 

In those spots we'd need to use the critical section API. We may also be able to expand usage of `NPY_BEGIN_THREADS_DESCR` `NPY_END_THREADS_DESCR` to handle this without adding a bunch of new boilerplate. The one sticking point that comes to mind is the need to have a reference to the array object, and the `ArrayMethod` API intentionally does not expose access to the array object. I'm not sure yet if that's a problem.

However all that said, I think given the release timing we might just need to document this limitation for the 2.1 release as a known issue. As far as I know this hasn't caused any crashes with pandas or scikit-learn, and they heavily use object string arrays.

**Comment by albanD** [comment]
Building latest 3.13 branch of cpython at f7b2b2a8fb167ee259990bd837db81d5f2b8d5a6 (July 17th) and installing numpy with `pip install --pre -i https://pypi.anaconda.org/scientific-python-nightly-wheels/simple numpy` (giving me 2.1.0.dev0), I get a segfault when trying to `import numpy.random`.

The stack trace is as follows:
```
Thread 1 "python" received signal SIGSEGV, Segmentation fault.
0x00007fffe99d4728 in __Pyx__ExceptionSave.isra.0 ()
   from /home/albandes/local/pytorch/3.13.0a6_nogil_source_env/lib/python3.13t/site-packages/numpy/random/_common.cpython-313t-x86_64-linux-gnu.so
(gdb) bt
#0  0x00007fffe99d4728 in __Pyx__ExceptionSave.isra.0 ()
   from /home/albandes/local/pytorch/3.13.0a6_nogil_source_env/lib/python3.13t/site-packages/numpy/random/_common.cpython-313t-x86_64-linux-gnu.so
#1  0x00007fffe99d3190 in __pyx_pymod_exec__common ()
   from /home/albandes/local/pytorch/3.13.0a6_nogil_source_env/lib/python3.13t/site-packages/numpy/random/_common.cpython-313t-x86_64-linux-gnu.so
#2  0x0000000000500de2 in PyModule_ExecDef (module=module@entry=<module at remote 0x20001a03b90>, 
    def=def@entry=0x7fffe99fe800 <__pyx_moduledef>) at Objects/moduleobject.c:489
#3  0x0000000000613e56 in exec_builtin_or_dynamic (mod=mod@entry=<module at remote 0x20001a03b90>)
    at Python/import.c:808
#4  0x0000000000613e79 in _imp_exec_dynamic_impl (module=<optimized out>, 
    mod=mod@entry=<module at remote 0x20001a03b90>) at Python/import.c:4739
#5  0x0000000000614244 in _imp_exec_dynamic (module=<optimized out>, mod=mod@entry=<module at remote 0x20001a03b90>)
    at Python/clinic/import.c.h:513
#6  0x00000000004fe886 in cfunction_vectorcall_O (
    func=<built-in method exec_dynamic of module object at remote 0x20000392ee0>, args=<optimized out>, 
    nargsf=<optimized out>, kwnames=<optimized out>) at Objects/methodobject.c:512
#7  0x00000000004b3262 in _PyVectorcall_Call (tstate=tstate@entry=0x9bbb00 <_PyRuntime+417024>, 
    func=0x4fe81f <cfunction_vectorcall_O>, 
    callable=callable@entry=<built-in method exec_dynamic of module object at remote 0x20000392ee0>, 
    tuple=tuple@entry=(<module at remote 0x20001a03b90>,), kwargs=kwargs@entry={}) at Objects/call.c:273
...
```
With the python stack trace:
```
(gdb) py-bt
Traceback (most recent call first):
  <built-in method exec_dynamic of module object at remote 0x20000392ee0>
  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 1324, in exec_module
  File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  <built-in method __import__ of module object at remote 0x200003900f0>
  <built-in method exec_dynamic of module object at remote 0x20000392ee0>
  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 1324, in exec_module
  File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "/home/albandes/local/pytorch/3.13.0a6_nogil_source_env/lib/python3.13t/site-packages/numpy/random/_pickle.py", line 1, in <module>
    from .bit_generator import BitGenerator
 ...
```

**Comment by seberg** [comment]
> The one sticking point that comes to mind is the need to have a reference to the array object, and the ArrayMethod API intentionally does not expose access to the array object

You would have to walk the base anyway, which isn't super nice, of course you could add an *owner* field to the additional information that is passed in via the struct, though, if that ends up as the best solution.
I still think the StringDType design is sound to do this via the dtype instance, but without a side-car buffer I dunno if an explicit owner somewhere else might end up easier/cleaner.

There are still some issues and I don't know if they even *can* be solved fully currently:
* Even within NumPy you would have to walk the `base` which is a bit annoying currently (the dtype instance would solve this a bit).
* Memoryviews and other protocols allow sharing of Python objects but ("immutable" flag says hello), it has no mechanism to pass on the right "owner" a critical section should lock on.
* It seems the critical section API is limited to two objects, but more objects are possible:
  * Is there a way to fallback to something else?
  * One could potentially do defensive copies.  For non-fast paths, I think it would actually make sense to do this in the buffering layer (copying objects is pretty cheap and then you get away with only ever locking a single "owner" at the same time).

Unless there is a fallback like `Py_BEGIN_CRITICAL_SECTION(NULL)` to say that the owning/critical object is unknown?! (Practically re-enabling the GIL, though, so it seems mainly a good fallback for memory sharing.).

**Comment by EwoutH** [comment]
3.13 rc1 is released! This means the ABI is stable and wheels can be uploaded to PyPI.

Tracking for cibuildwheel: https://github.com/pypa/cibuildwheel/pull/1949

**Comment by ngoldbaum** [comment]
We're planning to upload `cp313t` wheels to pypi for Mac and Linux for the NumPy 2.1 release candidate. 

Hopefully we'll be able to get Windows wheels up too but that's waiting on resolving some tricky merge conflicts in our meson patches, which will also hopefully help us to upstream the meson patches themselves.

**Comment by EwoutH** [comment]
cibuildwheel now uses the latest Python `3.13.0rc1` and builds Python 3.13 by default (https://github.com/pypa/cibuildwheel/pull/1950). Next cibuildwheel release should be soon. 

Free-threaded wheels are still behind a flag ([CIBW_FREE_THREADED_SUPPORT](https://cibuildwheel.pypa.io/en/stable/options/#free-threaded-support)), but can also be build and uploaded using the ABI stable Python 3.13.0rc1.

**Comment by ngoldbaum** [comment]
Yup, right now the suggestion is: don't do that.

I'm going to add docs about this but NumPy isn't thread safe and has never been thread safe. Free-threading makes this more acute but it's a pre-existing problem.

My plan for NumPy 2.1 is to document that mutating shared state in multithreaded python code should be done with extreme care and may cause crashes if there are races. This is also true with the GIL. It's certainly easier to cause a crash in the free-threaded build but the same issues about shared mutable state happen if any threads release the GIL and touch shared state.

There are also probably some safety issues around object arrays that are only present in the GIL-disabled build.

We're going to need to think carefully about adding improved locking around ndarray itself. It'll be a big project.

**Comment by rgommers** [comment]
If anyone was waiting for Windows wheels: there are nightlies at https://anaconda.org/scientific-python-nightly-wheels/numpy/files now. The next release (2.1.3) will include them on PyPI.

**Comment by ngoldbaum** [comment]
NumPy 2.1 included initial support for free-threading, so I'm closing this issue (a bit belatedly).

The main followup is https://github.com/numpy/numpy/issues/27199, which I'm hoping to have time to work on soon.
